tests/framework/umbra_client_faulty_redzone.c

```diff
@@ -0,0 +1,262 @@
+
+#ifdef WINDOWS
+static bool event_exception_instrumentation(void *drcontext,
+                                            dr_exception_t *excpt);
+#else
+static dr_signal_action_t event_signal_instrumentation(void *drcontext,
+                                                       dr_siginfo_t *info);
+#endif
+
+static void exit_event(void);
+
+DR_EXPORT void dr_client_main(client_id_t id, int argc, const char *argv[]) {
+  drreg_options_t ops = {sizeof(ops), 4, true};
+  umbra_map_options_t umbra_map_ops;
+
+  drmgr_init();
+  drreg_init(&ops);
+
+  memset(&umbra_map_ops, 0, sizeof(umbra_map_ops));
+  umbra_map_ops.scale = UMBRA_MAP_SCALE_DOWN_4X;
+  umbra_map_ops.flags =
+      UMBRA_MAP_CREATE_SHADOW_ON_TOUCH | UMBRA_MAP_SHADOW_SHARED_READONLY;
+  umbra_map_ops.default_value = 0;
+  umbra_map_ops.default_value_size = 1;
+  /* Enable redzones and make them faulty. */
+  umbra_map_ops.make_redzone_faulty = true;
+  if (umbra_init(id) != DRMF_SUCCESS)
+    DR_ASSERT_MSG(false, "fail to init umbra");
+  if (umbra_create_mapping(&umbra_map_ops, &umbra_map) != DRMF_SUCCESS)
+    DR_ASSERT_MSG(false, "fail to create shadow memory mapping");
+  drmgr_register_bb_instrumentation_event(NULL, event_app_instruction, NULL);
+#ifdef WINDOWS
+  drmgr_register_exception_event(event_exception_instrumentation);
+#else
+  drmgr_register_signal_event(event_signal_instrumentation);
+#endif
+  dr_register_exit_event(exit_event);
+}
+
+static void instrument_mem(void *drcontext, instrlist_t *ilist, instr_t *where,
+                           opnd_t ref) {
+  reg_id_t regaddr;
+  reg_id_t scratch;
+  bool ok;
+
+#ifdef X86
+  drvector_t allowed;
+  drreg_init_and_fill_vector(&allowed, false);
+  drreg_set_vector_entry(&allowed, DR_REG_XCX, true);
+#endif
+
+  if (drreg_reserve_aflags(drcontext, ilist, where) != DRREG_SUCCESS ||
+      drreg_reserve_register(drcontext, ilist, where,
+                             IF_X86_ELSE(&allowed, NULL),
+                             &scratch) != DRREG_SUCCESS ||
+      drreg_reserve_register(drcontext, ilist, where, NULL, &regaddr) !=
+          DRREG_SUCCESS) {
+    DR_ASSERT(false); /* Can't recover. */
+    return;
+  }
+#ifdef X86
+  drvector_delete(&allowed);
+#endif
+
+  ok = drutil_insert_get_mem_addr(drcontext, ilist, where, ref, regaddr,
+                                  scratch);
+  DR_ASSERT(ok);
+  /* Save the app address to a well-known spill slot, so that the fault handler
+   * can recover if no shadow memory was installed yet.
+   */
+  dr_save_reg(drcontext, ilist, where, regaddr, SPILL_SLOT_2);
+
+  if (umbra_insert_app_to_shadow(drcontext, umbra_map, ilist, where, regaddr,
+                                 &scratch, 1) != DRMF_SUCCESS) {
+    DR_ASSERT(false);
+  }
+  instrlist_meta_preinsert(
+      ilist, where,
+      INSTR_XL8(
+          XINST_CREATE_store_1byte(
+              drcontext,
+              OPND_CREATE_MEM8(regaddr, dr_page_size() /* enter redzone */),
```

Comment:
So this is deliberately *not* accessing the proper shadow xl8, but instead jumping a page ahead, for the sole purpose of testing the redzone, right?  Maybe worth a longer comment.

Reply:
Done.